src/Control/Effect/Diagnostics.hs

```diff
@@ -83,6 +85,14 @@ recover = fmap (either (const Nothing) Just) . recover'
 recover' :: Has Diagnostics sig m => m a -> m (Either SomeDiagnostic a)
 recover' = send . Recover'
 
+-- | Form an "error boundary" around an action, where:
+-- - errors and warnings cannot "escape" the scope of the action
+-- - warnings from outside of the error boundary do not impact the FailureBundles produced by the action
+--
+-- This returns a FailureBundle if the action failed; otherwise returns the result of the action
+errorBoundary :: Has Diagnostics sig m => m a -> m (Either FailureBundle a)
```

Comment:
Agreed. I'll put together some docs. `errorBoundary` is primarily a hack, and is otherwise analogous to `runDiagnostics`.

`errorBoundary` is a workaround to allow us to send the `Diagnostics` constraint upward, so we can intercept `context` calls from the inner action. We use this for `stickyDiag` specifically